File: prisma/seed.ts
import fs from "fs";
import path from "path";
import csv from "csv-parser";
import { prisma } from "../lib/prisma";
import { LacunaCategory } from "@prisma/client";

function readCSV(fileName: string): Promise<any[]> {
  const filePath = path.join(process.cwd(), "prisma/data", fileName);
  return new Promise((resolve, reject) => {
    const rows: any[] = [];
    fs.createReadStream(filePath)
      .pipe(csv())
      .on("data", (row) => rows.push(row))
      .on("end", () => resolve(rows))
      .on("error", reject);
  });
}

async function seedVirtues() {
  const rows = await readCSV("virtues.csv");

  for (const r of rows) {
    await prisma.virtue.upsert({
      where: { nameEn: r.name_en },
      update: { nameMr: r.name_mr },
      create: {
        nameEn: r.name_en,
        nameMr: r.name_mr,
      },
    });
  }
}

async function seedSubVirtues() {
  const rows = await readCSV("subvirtues.csv");

  for (const r of rows) {
    const virtue = await prisma.virtue.findUnique({
      where: { nameEn: r.virtue_name_en },
    });

    if (!virtue) {
      throw new Error(`Virtue not found: ${r.virtue_name_en}`);
    }

    await prisma.subVirtue.upsert({
      where: { nameEn: r.name_en },
      update: {
        nameMr: r.name_mr,
        virtueId: virtue.id,
      },
      create: {
        nameEn: r.name_en,
        nameMr: r.name_mr,
        virtueId: virtue.id,
      },
    });
  }
}

async function seedSentences() {
  const rows = await readCSV("sentences.csv");

  for (const r of rows) {
    const subVirtue = await prisma.subVirtue.findUnique({
      where: { nameEn: r.subvirtue_name_en },
    });

    if (!subVirtue) {
      throw new Error(`SubVirtue not found: ${r.subvirtue_name_en}`);
    }

    await prisma.sentence.upsert({
      where: { textEn: r.text_en },
      update: {
        textMr: r.text_mr,
        subVirtueId: subVirtue.id,
      },
      create: {
        textEn: r.text_en,
        textMr: r.text_mr,
        subVirtueId: subVirtue.id,
      },
    });
  }
}

async function seedLacunae() {
  const rows = await readCSV("lacunae.csv");

  for (const r of rows) {
    const category = r.category as LacunaCategory;

    if (!["A", "B", "C"].includes(category)) {
      throw new Error(
        `Invalid category for lacuna ${r.name_en}: ${r.category}`
      );
    }

    await prisma.lacuna.upsert({
      where: { nameEn: r.name_en },
      update: {
        nameMr: r.name_mr,
        category,
      },
      create: {
        nameEn: r.name_en,
        nameMr: r.name_mr,
        category,
      },
    });
  }
}


async function seedLacunaSubVirtues() {
  const rows = await readCSV("lacuna_subvirtues.csv");

  for (const r of rows) 
    {

    console.log(
      "Mapping row:",
      r.lacuna_name_en,
      "→",
      JSON.stringify(r.subvirtue_name_en)
    );

    const lacuna = await prisma.lacuna.findUnique({
      where: { nameEn: r.lacuna_name_en },
    });
    const subVirtue = await prisma.subVirtue.findUnique({
      where: { nameEn: r.subvirtue_name_en },
    });

    if (!lacuna || !subVirtue) {
      throw new Error(
        `Mapping failed: ${r.lacuna_name_en} → ${r.subvirtue_name_en}`
      );
    }

    await prisma.lacunaSubVirtue.upsert({
      where: {
        lacunaId_subVirtueId: {
          lacunaId: lacuna.id,
          subVirtueId: subVirtue.id,
        },
      },
      update: { priority: Number(r.priority) },
      create: {
        lacunaId: lacuna.id,
        subVirtueId: subVirtue.id,
        priority: Number(r.priority),
      },
    });
  }
}

async function main() {
  console.log("🌱 Veervrat ontology seeding started");

  await seedVirtues();
  await seedSubVirtues();
  await seedSentences();
  await seedLacunae();
  await seedLacunaSubVirtues();

  console.log("✅ Veervrat ontology seeding completed");
}

main()
  .then(async () => {
    await prisma.$disconnect();
  })
  .catch(async (e) => {
    console.error(e);
    await prisma.$disconnect();
    process.exit(1);
  });

File: backend/app/seed/seed.py
"""
Seed script: reads CSV data from prisma/data/ and inserts into the new database.
Run from backend/ directory:
    python -m app.seed.seed
"""
import csv
import os
import sys
import uuid

# Ensure project root is discoverable
sys.path.insert(0, os.path.dirname(os.path.dirname(os.path.dirname(os.path.abspath(__file__)))))

from app.database import SessionLocal
from app.models.models import (
    Lacuna, Virtue, SubVirtue, LacunaSubVirtue, Sentence, LacunaCategory,
    ExposureCatalogItem, ResolutionCatalogItem, ChallengeCatalogItem,
)

DATA_DIR = os.path.join(
    os.path.dirname(os.path.abspath(__file__)),
    "../../../prisma/data"
)


def seed():
    db = SessionLocal()
    try:
        # Check if already seeded
        if db.query(Lacuna).count() > 0:
            print("Database already seeded. Skipping.")
            return

        print("Seeding database...")

        # ── Virtues ──────────────────────────────────────────────────────
        virtue_map: dict[str, str] = {}  # nameEn -> id
        with open(os.path.join(DATA_DIR, "virtues.csv"), encoding="utf-8") as f:
            for row in csv.DictReader(f):
                v_id = str(uuid.uuid4())
                virtue = Virtue(id=v_id, name_en=row["name_en"], name_mr=row["name_mr"])
                db.add(virtue)
                virtue_map[row["name_en"]] = v_id
        db.flush()
        print(f"  Seeded {len(virtue_map)} virtues")

        # ── SubVirtues ────────────────────────────────────────────────────
        sv_map: dict[str, str] = {}  # nameEn -> id
        with open(os.path.join(DATA_DIR, "subvirtues.csv"), encoding="utf-8") as f:
            for row in csv.DictReader(f):
                virtue_id = virtue_map.get(row["virtue_name_en"])
                if not virtue_id:
                    print(f"  WARNING: virtue '{row['virtue_name_en']}' not found for subvirtue '{row['name_en']}'")
                    continue
                sv_id = str(uuid.uuid4())
                sv = SubVirtue(id=sv_id, name_en=row["name_en"], name_mr=row["name_mr"], virtue_id=virtue_id)
                db.add(sv)
                sv_map[row["name_en"]] = sv_id
        db.flush()
        print(f"  Seeded {len(sv_map)} sub-virtues")

        # ── Lacunae ───────────────────────────────────────────────────────
        lacuna_map: dict[str, str] = {}  # nameEn -> id
        with open(os.path.join(DATA_DIR, "lacunae.csv"), encoding="utf-8") as f:
            for row in csv.DictReader(f):
                l_id = str(uuid.uuid4())
                cat = LacunaCategory(row["category"].strip().upper())
                lacuna = Lacuna(id=l_id, name_en=row["name_en"], name_mr=row["name_mr"], category=cat)
                db.add(lacuna)
                lacuna_map[row["name_en"]] = l_id
        db.flush()
        print(f"  Seeded {len(lacuna_map)} lacunae")

        # ── LacunaSubVirtues ──────────────────────────────────────────────
        lsv_count = 0
        with open(os.path.join(DATA_DIR, "lacuna_subvirtues.csv"), encoding="utf-8") as f:
            for row in csv.DictReader(f):
                lacuna_id = lacuna_map.get(row["lacuna_name_en"])
                sv_id = sv_map.get(row["subvirtue_name_en"])
                if not lacuna_id or not sv_id:
                    continue
                lsv = LacunaSubVirtue(
                    id=str(uuid.uuid4()),
                    lacuna_id=lacuna_id,
                    sub_virtue_id=sv_id,
                    priority=int(row["priority"]),
                )
                db.add(lsv)
                lsv_count += 1
        db.flush()
        print(f"  Seeded {lsv_count} lacuna-subvirtue mappings")

        # ── Sentences ─────────────────────────────────────────────────────
        sentence_count = 0
        with open(os.path.join(DATA_DIR, "sentences.csv"), encoding="utf-8") as f:
            for row in csv.DictReader(f):
                sv_id = sv_map.get(row["subvirtue_name_en"])
                if not sv_id:
                    continue
                s = Sentence(
                    id=str(uuid.uuid4()),
                    text_en=row["text_en"],
                    text_mr=row["text_mr"],
                    sub_virtue_id=sv_id,
                )
                db.add(s)
                sentence_count += 1
        db.flush()
        print(f"  Seeded {sentence_count} sentences")

        db.commit()
        print("✓ Seeding complete!")

    except Exception as e:
        db.rollback()
        print(f"Error during seeding: {e}")
        raise
    finally:
        db.close()


def seed_catalog():
    """
    Seed catalog items for each sentence.
    Safe to run multiple times — skips if already populated.
    """
    db = SessionLocal()
    try:
        if db.query(ExposureCatalogItem).count() > 0:
            print("Catalog already seeded. Skipping.")
            return

        sentences = db.query(Sentence).all()
        if not sentences:
            print("No sentences found — run seed() first.")
            return

        print(f"Seeding catalog for {len(sentences)} sentences...")

        exposure_templates = [
            ("Read about it", "Find and read an article, book chapter, or resource related to this virtue."),
            ("Observe in real life", "Notice examples of this virtue (or its absence) in people around you today."),
            ("Discuss with someone", "Talk to a trusted person about what this virtue means and where you see it practiced."),
        ]
        resolution_templates = [
            ("Morning intention", "Start each day by setting a clear intention to practice this virtue.", "Daily"),
            ("Evening reflection", "Before bed, review one moment from the day where this virtue showed up or was missing.", "Daily"),
            ("Weekly journal entry", "Write a short journal entry about your progress with this virtue each week.", "Weekly"),
        ]
        challenge_templates = [
            (
                "Demonstrate it in a real situation",
                "Identify one upcoming real-life situation where you can consciously apply this virtue.",
                "Act on this virtue in a real situation at least once, then reflect on what happened.",
            ),
            (
                "Teach or share it",
                "Explain this virtue and why it matters to someone in your life.",
                "Have a genuine conversation where you share what you've learned about this virtue.",
            ),
        ]

        for sentence in sentences:
            for title, description in exposure_templates:
                db.add(ExposureCatalogItem(
                    id=str(uuid.uuid4()),
                    sentence_id=sentence.id,
                    title=title,
                    description=description,
                ))
            for title, description, freq_hint in resolution_templates:
                db.add(ResolutionCatalogItem(
                    id=str(uuid.uuid4()),
                    sentence_id=sentence.id,
                    title=title,
                    description=description,
                    frequency_hint=freq_hint,
                ))
            for title, description, criteria in challenge_templates:
                db.add(ChallengeCatalogItem(
                    id=str(uuid.uuid4()),
                    sentence_id=sentence.id,
                    title=title,
                    description=description,
                    achievement_criteria=criteria,
                ))

        db.commit()
        total = len(sentences) * (len(exposure_templates) + len(resolution_templates) + len(challenge_templates))
        print(f"✓ Catalog seeding complete! ({total} items)")

    except Exception as e:
        db.rollback()
        print(f"Error during catalog seeding: {e}")
        raise
    finally:
        db.close()


if __name__ == "__main__":
    seed()
    seed_catalog()
